pulsejet/nextcloud-oidc-login
# Add WebDAV access using OIDC bearer token

This PR adds the ability to allow WebDAV access using an OIDC bearer token. This can be useful when using [rclone WebDAV with OIDC](https://rclone.org/webdav/#openid-connect) to synchronize a users files from Nextcloud.

To avoid too much load on the OIDC provider, this PR also adds caching of the JWKs as well as the discovered OIDC configuration for a configurable amount of time.

I suggest reviewing this PR by going through the commits. I tried to make each commit a logical unit of change to aid in the review. 

## Testing

You can use the following script to test this functionality. To pass the tokens `audience` check in the `OpenIDConnectClient.php`, you will need to configure an audience mapper in Keycloak to ensure that your client is included in the `aud` property of the JWT. A [hardcoded audience mapper](https://www.keycloak.org/docs/latest/server_admin/#_audience_hardcoded) is sufficient here. Basically follow the following steps:

1. Go to `Client Scopes`
2. Add new client scope, call it `nextcloud`.
3. Under `Mappers` create a new mapper of type `Audience` and ensure that `Included Client Audience` contains your Nextcloud client ID. Click Save.
4. Finally, go to `Client > your-nextcloud-client > Client Scopes` and add the new `nextcloud` scope.

```bash
#!/bin/bash

KC_BASE="<keycloak-base>" # e.g. http://localhost:8080/auth
KC_REALM="<your-keycloak-realm>"
KC_USERNAME="<keycloak-username>"
KC_PASSWORD="<keycloak-password>"
KC_CLIENT="<your-nextcloud-client>"
KC_SECRET="<your-nextcloud-secret>"

NC_BASE="<nextcloud-base>" # e.g. http://localhost:8081
NC_UID="<nextcloud-user-id>"

resp=$(curl -s -X POST "$KC_BASE/realms/$KC_REALM/protocol/openid-connect/token" \
 -H "Content-Type: application/x-www-form-urlencoded" \
 -d 'grant_type=password' \
 -d "client_id=$KC_CLIENT" -d "client_secret=$KC_SECRET" \
 -d "username=$KC_USERNAME" -d "password=$KC_PASSWORD" \
 -d "scope=nextcloud")

token=$(echo $resp | jq -r .access_token)

[[ ! -n "$token" || "$token" == "null" ]]; then
    echo $resp
    exit 1
fi

curl -X PROPFIND \
    -H "Depth: 1" \
    -H "Authorization: Bearer $token" \
    $NC_BASE/remote.php/dav/files/$NC_UID/
```

If everything worked as expected, you should get a long XML response that looks something like this:

```xml
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns" xmlns:nc="http://nextcloud.org/ns">
  <d:response>
    <d:href>/remote.php/dav/files/4b5e7a7e-6fd9-4c5a-b4be-18bbfd4eed64/</d:href>
    <d:propstat>
      <d:prop>
        <d:getlastmodified>Sat, 24 Apr 2021 13:50:01 GMT</d:getlastmodified>
        <d:resourcetype>
          <d:collection />
        </d:resourcetype>
        <d:quota-used-bytes>16611184</d:quota-used-bytes>
        <d:quota-available-bytes>-3</d:quota-available-bytes>
        <d:getetag>"60842209bab96"</d:getetag>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
</d:multistatus>
```
